tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,565 @@
+			}
+			blob, err := w.cfg.ProofArchive.FetchProof(ctx, locator)
+			if err != nil {
+				startErr = fmt.Errorf("unable to fetch proof "+
+					"for asset %v: %w", assets[idx].ID(),
+					err)
+				return
+			}
+
+			f := &proof.File{}
+			err = f.Decode(bytes.NewReader(blob))
+			if err != nil {
+				startErr = fmt.Errorf("unable to decode "+
+					"proof: %w", err)
+				return
+			}
+
+			err = w.MaybeWatch(f, w.DefaultUpdateCallback())
+			if err != nil {
+				startErr = fmt.Errorf("unable to watch proof "+
+					"for asset %v: %w", assets[idx].ID(),
+					err)
+				return
+			}
+		}
+	})
+	return startErr
+}
+
+// Stop signals for a re-org watcher to gracefully exit.
+func (w *ReOrgWatcher) Stop() error {
+	var stopErr error
+	w.stopOnce.Do(func() {
+		log.Info("Stopping re-org watcher")
+
+		close(w.Quit)
+		w.Wg.Wait()
+	})
+
+	return stopErr
+}
+
+// waitForConf waits for the anchor transaction of the given proofs to reach a
+// safe confirmation depth.
+func (w *ReOrgWatcher) waitForConf(ctx context.Context, txHash chainhash.Hash,
+	newProofs *proofRegistration) error {
+
+	// Do we already have a confirmation watcher for this transaction? Then
+	// we don't need to add another one.
+	existingNtfn, ok := w.pendingProofs[txHash]
+	if ok {
+		existingNtfn.proofsRegistrations = append(
+			existingNtfn.proofsRegistrations, newProofs,
+		)
+
+		return nil
+	}
+
+	reOrgChan := make(chan struct{}, 1)
+	ctxc, cancel := context.WithCancel(ctx)
+
+	// We only register for a single confirmation, which we expect to come
+	// in immediately, since we're only given the proof after one conf.
+	// But we keep the registration open, so we will get notified again if
+	// there is a re-org.
+	confEvent, errChan, err := w.cfg.ChainBridge.RegisterConfirmationsNtfn(
+		ctxc, &txHash, newProofs.anchorTx.TxOut[0].PkScript, 1,
+		uint32(newProofs.blockHeight), true, reOrgChan,
+	)
+	if err != nil {
+		cancel()
+		return fmt.Errorf("unable to register for conf ntfn: %w", err)
+	}
+
+	// Now that we have created the confirmation watcher, we can add the
+	// new proofs to the pending proofs map.
+	w.pendingProofs[txHash] = &anchorTxNotification{
+		proofsRegistrations: []*proofRegistration{newProofs},
+		cancel:              cancel,
+	}
+
+	w.Wg.Add(1)
+	go func() {
+		defer confEvent.Cancel()
+		defer w.Wg.Done()
+		defer cancel()
+
+		for {
+			select {
+			// If the transaction was confirmed normally, we don't
+			// need to do anything and can stop watching it.
+			case conf, ok := <-confEvent.Confirmed:
+				if !ok {
+					return
+				}
+
+				select {
+				case w.incomingConfs <- conf:
+				case <-w.Quit:
+				}
+
+			// If the transaction was re-organized out, we need to
+			// update the proof and store it in the proof archive.
+			case <-reOrgChan:
+				log.Infof("Anchor TX %v was re-organized out "+
+					"of the chain, will update proof "+
+					"with next confirmation", txHash)
+
+				// We continue to watch the transaction until
```

Comment:
Ahh ok I think that's the behavior I missed before: that we'll get another confirmation notification in the event of a re-org. So given that, why do we even need to reorg channel in the first place?

Reply:
For now, just for logging, which was helpful for debugging the itest. In the future, we could react to it.